engine/uieventmanager.cpp

```diff
@@ -0,0 +1,50 @@
+#include "uieventmanager.h"
+
+#include "button.h"
+#include "colliderbox.h"
+#include "logger.h"
+#include "input.h"
+#include "uielement.h"
+
+namespace pong
+{
+
+void UIEventManager::ProcessEvents(const std::vector<std::unique_ptr<UIElement>>& uiElements)
+{
+    const glm::vec3 mousePosition = Input::GetMousePosition();
+
+    for (auto& uiElement : uiElements)
+    {
+        if (uiElement->GetType() != UIElementType::Button)
+        {
+            continue;
+        }
+
+        auto& button = static_cast<Button&>(*uiElement);
+        if (button.GetColliderBox()->CheckPointInBounds(mousePosition))
+        {
+            if (!button.IsHovered())
+            {
+                button.OnEvent(ButtonEvent::Hover);
+            }
+
+            if (Input::IsMouseButtonPressed(GLFW_MOUSE_BUTTON_LEFT))
+            {
+                if (!button.IsPressed())
+                {
+                    button.OnEvent(ButtonEvent::Pressed);
+                }
+            }
+            else if (button.IsPressed())
+            {
+                button.OnEvent(ButtonEvent::Release);
```

Comment:
Is this correct, on button pressed send event released?

Reply:
Yes, although that is confusing. This is similar to that collision detection system I have. If they are just now colliding call Enter, if they already were call Stay(), if they were but ar no longer call Stop(), I renamed this to WasPressed so it more clear that it was being pressed and no longer is, so release should be called.